0000-frontend-development-infrastructure-and-workflow-changes.md

```diff
@@ -0,0 +1,75 @@
+- Feature Name: frontend-development-infrastructure-and-workflow-changes
+- Start Date: 2016-07-20
+- RFC PR: 
+- Issue: 
+
+# Summary
+[summary]: #summary
+
+Currently, client side javascript in foreman is implemented with native javascript and the jquery library and managed by the Rails asset pipeline. In addition, various unrelated libraries are used for specific tasks, charting for example. This RFC proposes adding the React javascript library to the technology stack via NPM and Webpack. [(see Webpack PR)](https://github.com/theforeman/foreman/pull/3433) 
+
+# Motivation
+[motivation]: #motivation
+
+The current setup is based on software which was introduced many years ago, solved old problems that are no longer relevant and do not address the challenges that frontend developers face today. Many of the libraries we use are no longer maintained. The infrastructure is sparse and the responsibility for enforcing best practices and coding standards falls on the developer.
+ 
+Frontend developers today are facing much more ambitious challenges in implementing complex features, a trend that has skyrocketed with the move to ajax and client side rendering.
+
+The frontend ecosystem has expanded dramatically with the addition of many libraries and frameworks, as well as developer productivity tools, to meet these challenges. Reactjs is among these additions.
+
+Reactjs is a view library that falls under the umbrella MV* libraries and frameworks. The common thread among these frameworks is the separation of the management of state (model) from the process of rendering HTML (view). The benefits are similar to those provided by server side MVC, such as Rails.
+
+Applying an MV* approach would facilitate enforcement of best practices and coding standards and make the code produced more of robust, maintainable and extensible. It must be noted however that, any solution must 'work well' with the current foreman Ruby/Rails code base and plugins, as there is no intention to abandon it.
+
+The tools in the MV* space are often classified as being opinionated or unopinionated with respect to the amount of flexibility they allow the developer. This often maps to the breadth of the issues the library or framework attempts to deal with.
+
+Reactjs is a component-based rendering engine. As such, it is both narrow in scope and unopinionated. This will facilitate integration with Rails with minimum side effects on the existing Ruby/Rails code base compared to the opinionated alternatives.
+
+React components demonstrate a high level of encapsulation and support dependency injection. As a result, the components are very testable. The implementation of the recommendations in this RFC will be accompanied by the addition of linting and testing to the client side development process.
```

Comment:
@dLobatog faster development experience with TDD and not having to run Rails at all.